lib/phoenix/endpoint/socket_path_helpers.ex

```diff
@@ -0,0 +1,67 @@
+defmodule Phoenix.Endpoint.SocketPathHelpers do
+  @moduledoc ~S"""
+  Helper functions to build matcher paths & socket configuration, shared between Phoenix.Endpoint
+  (who needs to configure long polling & legacy Cowboy websocket handlers) and Phoenix.Endpoint.Sock
+
+  Only used at compile time, in the context of `__before_compile__` macros
+  """
+
+  def socket_paths(endpoint, path, socket, opts, module) do
+    websocket_paths(path, socket, opts, module) ++
+      longpoll_paths(endpoint, path, socket, opts, module)
+  end
+
+  def websocket_paths(path, socket, opts, module) do
+    websocket = Keyword.get(opts, :websocket, true)
+
+    if websocket do
+      config = Phoenix.Socket.Transport.load_config(websocket, Phoenix.Transports.WebSocket)
+      {conn_ast, match_path} = socket_path(path, config, module)
+      [{match_path, :websocket, conn_ast, socket, config}]
+    else
+      []
+    end
+  end
+
+  def longpoll_paths(endpoint, path, socket, opts, module) do
+    longpoll = Keyword.get(opts, :longpoll, false)
+
+    if longpoll do
+      config = Phoenix.Socket.Transport.load_config(longpoll, Phoenix.Transports.LongPoll)
+      plug_init = {endpoint, socket, config}
+      {conn_ast, match_path} = socket_path(path, config, module)
+      [{match_path, :plug, conn_ast, Phoenix.Transports.LongPoll, plug_init}]
+    else
+      []
+    end
+  end
+
+  defp socket_path(path, config, module) do
+    end_path_fragment = Keyword.fetch!(config, :path)
+
+    {vars, path} =
+      String.split(path <> "/" <> end_path_fragment, "/", trim: true)
+      |> Enum.join("/")
+      |> Plug.Router.Utils.build_path_match()
+
+    conn_ast =
+      if vars == [] do
+        quote context: module do
```

Comment:
`module` here is the eventual context in which this AST will be applied. I don't believe that the usual `Macro.var(:conn, nil)` approach will work here, since the final point-of-use for this intends for `conn` to bind against a variable defined in the containing macro itself (see [lib/phoenix/endpoint/sock.ex:165](https://github.com/phoenixframework/phoenix/pull/4973/files#diff-9f7279f1d916a972e876499435496ca3c1ea0aa688cc7228ce1e3ba4043f5994R165)). I'd love to be proven wrong here, but 'functions used to build quotes which are later included in other macros across module boundaries' is rather at the outer limits of my meta programming universe :). 

A similar case exists on line 59, just below.